### PR TITLE
tikv/resolvelock: resolve lock in a batch

### PR DESCRIPTION
**Pull request by atmzhou**
This tikv resolves lock in a batch.

### CODE REVIEW

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -475,18 +476,25 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
             .with_label_values(&[label])
             .start_coarse_timer();
 
-        let commit_ts = match req.get_commit_version() {
-            0 => None,
-            x => Some(x),
-        };
+        let mut txn_2_status : HashMap<u64, u64> = HashMap::new();
```
Comment (siddontang) [inline src/server/service/kv.rs]: txn_status

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -884,52 +880,50 @@ fn process_write_impl(
...
-            for k in keys {
-                match commit_ts {
-                    Some(ts) => txn.commit(k, ts)?,
-                    None => txn.rollback(k)?,
-                }
-                if txn.write_size() >= MAX_TXN_WRITE_SIZE {
-                    scan_key = Some(k.to_owned());
+            let mut modifies : Vec<Modify> = vec![];
+            let rows = key_locks.len();
+            for key_lock in key_locks {
```
Comment (siddontang) [inline src/storage/txn/scheduler.rs]: use `for (x, y) in key_lock` to identify the tuple

**Comment by siddontang** [comment]
do we have a test to resolve many locks in a batch?

File: src/server/service/kv.rs [hunk 3]
```diff
@@ -16,6 +16,7 @@ use std::fmt::Debug;
 use std::io::Write;
 use std::sync::Arc;
 use std::sync::atomic::{AtomicUsize, Ordering};
+use std::collections::HashMap;
```
Comment (overvenus) [inline src/server/service/kv.rs]: use util::collections::HashMap;

File: src/server/service/kv.rs (hunk 1 above)
Comment (overvenus) [inline src/server/service/kv.rs]: Please run `make format`.

File: src/storage/txn/scheduler.rs [hunk 4]
```diff
@@ -1371,10 +1365,12 @@ impl Scheduler {
         to_be_write: Vec<Modify>,
         rows: usize,
     ) {
+	info!("we are going to write");
```
Comment (overvenus) [inline src/storage/txn/scheduler.rs]: Please make it `debug!` and add more details.

File: src/storage/txn/scheduler.rs [hunk 5]
```diff
@@ -1394,6 +1390,7 @@ impl Scheduler {
             .with_label_values(&[self.get_ctx_tag(cid), "write_finish"])
             .inc();
         debug!("write finished for command, cid={}", cid);
+        info!("write finished for command, cid={}", cid);
```
Comment (overvenus) [inline src/storage/txn/scheduler.rs]: `debug!`

**Comment by ngaut** [comment]
Please clean up your commit messages.

**Comment by atmzhou** [comment]
@disksing

**Comment by arthurprs** [comment]
Is this in a hot path? If so it might be better to use HashMap from util::collections instead.

File: src/server/service/kv.rs [hunk 6]
```diff
@@ -12,6 +12,7 @@
 // limitations under the License.
 
 use std::boxed::FnBox;
+use std::collections::HashMap;
```
Comment (siddontang) [inline src/server/service/kv.rs]: use util HashMap

File: src/server/service/kv.rs [hunk 7]
```diff
@@ -475,14 +476,21 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
             .with_label_values(&[label])
             .start_coarse_timer();
 
-        let commit_ts = match req.get_commit_version() {
-            0 => None,
-            x => Some(x),
-        };
+        let mut txn_2_status: HashMap<u64, u64> = HashMap::new();
```
Comment (siddontang) [inline src/server/service/kv.rs]: s/txn_2_status/txn_status/

File: src/storage/mod.rs [hunk 8]
```diff
@@ -12,6 +12,7 @@
 // limitations under the License.
 
 use std::thread;
+use std::collections::HashMap;
```
Comment (siddontang) [inline src/storage/mod.rs]: ditto

File: src/storage/mod.rs [hunk 9]
```diff
@@ -131,10 +133,9 @@ pub enum Command {
     ScanLock { ctx: Context, max_ts: u64 },
     ResolveLock {
         ctx: Context,
-        start_ts: u64,
-        commit_ts: Option<u64>,
+        txn2status: HashMap<u64, u64>,
```
Comment (siddontang) [inline src/storage/mod.rs]: s/txn2status/txn_status/

File: src/server/service/kv.rs [hunk 10]
```diff
@@ -475,14 +476,21 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
...
+
+        let start_ts = req.get_start_version();
+        if start_ts > 0 {
+            let commit_ts = req.get_commit_version();
+            txn_2_status.insert(start_ts, commit_ts);
+        } else {
+            for temp in req.take_txn_infos().into_iter() {
```
Comment (disksing) [inline src/server/service/kv.rs]: temp -> info.

File: tests/storage/sync_storage.rs [hunk 11]
```diff
@@ -150,9 +151,15 @@ impl SyncStorage {
     }
 
     pub fn resolve_lock(&self, ctx: Context, start_ts: u64, commit_ts: Option<u64>) -> Result<()> {
+        let mut temp_map = HashMap::new();
```
Comment (disksing) [inline tests/storage/sync_storage.rs]: temp_map -> txn_status

File: tests/storage/sync_storage.rs [hunk 12]
```diff
@@ -150,9 +151,15 @@ impl SyncStorage {
...
+        let temp = match commit_ts {
+            None => 0,
+            Some(x) => x,
+        };
+        temp_map.insert(start_ts, temp);
```
Comment (disksing) [inline tests/storage/sync_storage.rs]: temp_map.insert(start_ts, commit_ts.unwrap_or(0));

File: tests/storage/sync_storage.rs [hunk 13]
```diff
@@ -150,9 +151,15 @@ impl SyncStorage {
...
         wait_op!(|cb| {
             self.store
-                .async_resolve_lock(ctx, start_ts, commit_ts, cb)
+                .async_resolve_lock(ctx, temp_map.clone(), cb)
```
Comment (disksing) [inline tests/storage/sync_storage.rs]: seems don't need clone here.

File: src/storage/txn/scheduler.rs [hunk 14]
```diff
@@ -1375,6 +1375,7 @@ impl Scheduler {
             .with_label_values(&[self.get_ctx_tag(cid), "write"])
             .inc();
         if to_be_write.is_empty() {
+            info!("modifies is empty");
```
Comment (disksing) [inline src/storage/txn/scheduler.rs]: Why add a log here?

File: src/storage/txn/scheduler.rs [hunk 15]
```diff
@@ -605,24 +602,23 @@ fn process_read(
...
-                    if keys.is_empty() {
+                        .observe(key_locks.len() as f64);
+                    if key_locks.is_empty() {
                         Ok(None)
                     } else {
                         Ok(Some(Command::ResolveLock {
                             ctx: ctx.clone(),
-                            start_ts: start_ts,
-                            commit_ts: commit_ts,
+                            txn_status: txn_status.clone(),
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: Can we take `txn_status` away?

File: src/storage/txn/scheduler.rs [hunk 16]
```diff
@@ -884,52 +880,56 @@ fn process_write_impl(
...
+                        });
+                    }
+                    txn.commit(current_key, ts)?;
+                } else {
+                    txn.rollback(current_key)?;
                 }
-                if txn.write_size() >= MAX_TXN_WRITE_SIZE {
-                    scan_key = Some(k.to_owned());
+                modifies.append(&mut txn.modifies());
+                if modifies.len() >= MAX_TXN_WRITE_SIZE {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: Not length but bytes.

File: src/storage/txn/scheduler.rs [hunk 17]
```diff
@@ -884,52 +880,56 @@ fn process_write_impl(
...
+                let mut txn = MvccTxn::new(
+                    snapshot,
+                    statistics,
+                    current_lock.ts,
+                    None,
+                    ctx.get_isolation_level(),
+                    !ctx.get_not_fill_cache(),
+                );
+                let status = txn_status.get(&current_lock.ts);
+                let ts = match status {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: s/ts/commit_ts

File: src/storage/txn/scheduler.rs [hunk 18]
```diff
@@ -616,7 +616,7 @@ fn process_read(
                     } else {
                         Ok(Some(Command::ResolveLock {
                             ctx: ctx.clone(),
-                            txn_status: txn_status.clone(),
+                            txn_status: txn_status.drain().collect(),
```
Comment (arthurprs) [inline src/storage/txn/scheduler.rs]: If the original txn_status reused after this is called? If not the drain().collect() can become `mem::replace(txn_status, Default::default())`

Comment (siddontang) [reply src/storage/txn/scheduler.rs]: Thanks @arthurprs 

Seem we also use drain() + collect() in some places, could you help us to improve them?

File: src/storage/txn/scheduler.rs [hunk 19]
```diff
@@ -884,52 +881,58 @@ fn process_write_impl(
...
-            );
-            let rows = keys.len();
-            for k in keys {
-                match commit_ts {
-                    Some(ts) => txn.commit(k, ts)?,
-                    None => txn.rollback(k)?,
+            let mut modifies: Vec<Modify> = vec![];
+            let mut write_size = 0;
+            let rows = key_locks.len();
+            for &(ref current_key, ref current_lock) in key_locks {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: Can we group `key_locks` by `start_ts`?

File: src/storage/txn/scheduler.rs [hunk 20]
```diff
@@ -884,52 +881,58 @@ fn process_write_impl(
...
+                    None,
+                    ctx.get_isolation_level(),
+                    !ctx.get_not_fill_cache(),
+                );
+                let status = txn_status.get(&current_lock.ts);
+                let commit_ts = match status {
+                    Some(ts) => *ts,
+                    None => panic!("txn status not found!"),
+                };
+                if commit_ts > 0 {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: Please verify commit_ts is greater than start_ts.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
@AndreMouche PTAL

File: src/storage/txn/scheduler.rs [hunk 21]
```diff
@@ -884,52 +881,58 @@ fn process_write_impl(
...
+                    }
+                    txn.commit(current_key, commit_ts)?;
+                } else {
+                    txn.rollback(current_key)?;
                 }
-                if txn.write_size() >= MAX_TXN_WRITE_SIZE {
-                    scan_key = Some(k.to_owned());
+                write_size += txn.write_size();
+                modifies.append(&mut txn.modifies());
+                if write_size >= MAX_TXN_WRITE_SIZE {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: Please adjust `MAX_TXN_WRITE_SIZE` and `RESOLVE_LOCK_BATCH_SIZE` to make scan_lock never scan duplicated keys.

**Comment by siddontang** [comment]
PTAL @AndreMouche

File: src/server/service/kv.rs [hunk 22]
```diff
@@ -507,6 +515,7 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
             });
 
         ctx.spawn(future);
+
```
Comment (AndreMouche) [inline src/server/service/kv.rs]: remove empty line

File: src/storage/txn/scheduler.rs [hunk 23]
```diff
@@ -31,12 +31,14 @@
 //! is ensured by the transaction protocol implemented in the client library, which is transparent
 //! to the scheduler.
 
+use std::collections::HashMap;
```
Comment (AndreMouche) [inline src/storage/txn/scheduler.rs]: why not use `util::collections::HashMap` instead?

File: src/storage/txn/scheduler.rs [hunk 24]
```diff
@@ -884,52 +881,58 @@ fn process_write_impl(
...
+                    statistics,
+                    current_lock.ts,
+                    None,
+                    ctx.get_isolation_level(),
+                    !ctx.get_not_fill_cache(),
+                );
+                let status = txn_status.get(&current_lock.ts);
+                let commit_ts = match status {
+                    Some(ts) => *ts,
+                    None => panic!("txn status not found!"),
```
Comment (AndreMouche) [inline src/storage/txn/scheduler.rs]: .

File: src/storage/txn/scheduler.rs [hunk 25]
```diff
@@ -884,52 +881,58 @@ fn process_write_impl(
...
             if scan_key.is_none() {
-                (ProcessResult::Res, txn.modifies(), rows)
+                (ProcessResult::Res, modifies, rows)
             } else {
                 let pr = ProcessResult::NextCommand {
                     cmd: Command::ResolveLock {
                         ctx: ctx.clone(),
-                        start_ts: start_ts,
-                        commit_ts: commit_ts,
+                        txn_status: txn_status.clone(),
```
Comment (AndreMouche) [inline src/storage/txn/scheduler.rs]: could we move `txn_status` instead of `clone`?

File: tests/storage/sync_storage.rs [hunk 26]
```diff
@@ -150,10 +151,25 @@ impl SyncStorage {
...
-                .async_resolve_lock(ctx, start_ts, commit_ts, cb)
-                .unwrap()
+            self.store.async_resolve_lock(ctx, txn_status, cb).unwrap()
+        }).unwrap()
+    }
+
+    pub fn resolve_lock_batch(
+        &self,
+        ctx: Context,
+        start_ts: Vec<u64>,
```
Comment (AndreMouche) [inline tests/storage/sync_storage.rs]: It's a little weird to split `start_ts` and `commit_ts`, I perfer to use `Vec<(u64,u64)>` instead.

File: tests/storage/test_storage.rs [hunk 27]
```diff
@@ -392,6 +392,34 @@ fn test_txn_store_resolve_lock_batch(key_prefix_len: usize, n: usize) {
...
+            Mutation::Put((make_key(b"p2"), b"v10".to_vec())),
+            Mutation::Put((make_key(b"s2"), b"v10".to_vec())),
+        ],
+        b"p2",
+        10,
+    );
+    store.resolve_lock_batch_ok(5, 0, 10, 20);
+    store.get_none(b"p1", 20);
+    store.get_none(b"s1", 30);
+    store.get_ok(b"p2", 20, b"v10");
```
Comment (AndreMouche) [inline tests/storage/test_storage.rs]: I think the same `ts`  would be used in both `commit` and `read`, so may be a `ts` bigger than 20 would be more clear for the reviewer?

File: src/server/service/kv.rs [hunk 28]
```diff
@@ -475,14 +476,21 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
             .with_label_values(&[label])
             .start_coarse_timer();
 
-        let commit_ts = match req.get_commit_version() {
-            0 => None,
-            x => Some(x),
-        };
+        let mut txn_status = HashMap::default();
```
Comment (siddontang) [inline src/server/service/kv.rs]: with capacity

**Comment by disksing** [comment]
PTAL @siddontang

**Pull request merged**